ocr.js

```diff
@@ -1,14 +1,21 @@
 const tesseract = require('node-tesseract');
 
-var options = {
+function OCR(options) {
+  if (typeof options === "object") 
+    this.options = Object.assign(this.__proto__.options, options);
```

Comment:
Hi @gabrieldrs, I had a idea but I don't know if it is good... 
I tried the function and I saw that is assigning all the object that we pass to the new instance:
```
var ocrInstance5 = new OCR( {test: true});
console.log(ocrInstance5.options);

Output:
{ l: 'por', psm: 6, binary: '/usr/bin/tesseract', test: true }

```
I thought to change this line:

```
 this.options = Object.assign(this.__proto__.options, options.l, 
                                     options.psm, options.binary);

var ocrInstance5 = new OCR( {test: true});
console.log(ocrInstance5.options);
{ l: 'por', psm: 6, binary: '/usr/bin/tesseract' }

```
What do you think?

Reply:
Hey Gustavo, that's a great observation!

Let me try to explain this part of the code a little better:

The OCR object has a default value for `l`, `psm` and `binary`, these default values are used in case the caller hasn't specified them.

So, for example, when I run `var myOcr = new OCR({l: 'en'})`, `myOcr` will use the default values for both `psm` and `binary`, and my specified value for `l`. 

Your observation is completely right, I, as a caller, can specify any option I want, and this will be added to my options array, and this is an issue for sure.

The solution you suggested is almost correct, let me show you the results I got from it:

```js
var myOcr = new OCR({test:'a'})
console.log(myOcr.options)
// { l: 'por', psm: 6, binary: '/usr/bin/tesseract' }
myOcr = new OCR({l:'en'});
console.log(myOcr.options)
// { '0': 'e', '1': 'n', l: 'por', psm: 6, binary: '/usr/bin/tesseract' }
```
As you noticed when we specified a correct parameter, it had a somewhat "strange" behavior. Let me try to explain what happened: 

The [Object.assign()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) has the following signature: `Object.assign(target, ...sources)`, where target and sources are **objects**.
A string in javascript (and most other languages), is nothing more than an array of chars (char is a basic type that is 1 character long), and an array is an object (It's a little bit more complex than that, if you would like to understand this better, let me know).
In this case, `Object.assign` tried to assign the object `['e','n']` to the options object, creating two new indexes (0 and 1), and assigning the values 'e' and 'n' to them.


After your suggestion I started to think on a solution for the problem you pointed out and I came out with the following:

```js
function OCR(options){
  if (typeof options === "object") {
    const acceptedKeys = Object.keys(this.options);
    Object.keys(options).forEach((val) => {
      if (!acceptedKeys.includes(val))
        delete options[val];
    });
    this.options = Object.assign(this.__proto__.options, options);
  }
}
```
This is how it works:

1. I first create an array of accepted keys by getting a list of keys from `this.options`.
2. Then I get the list of keys defined by the caller
3. Then I run a function for each one of these keys checking if it's one of the accepted keys
4. In the case it is not, I remove it from the options object received
5. Finally, I assign the transformed object, containing only the accepted keys, to the instance's options object

And these are the results the above code generated:

```js
var myOCR = new OCR({test:'abc'})
console.log(myOCR.options);
// { l: 'por', psm: 6, binary: '/usr/bin/tesseract' }
var myOCR = new OCR({l:'en'})
console.log(myOCR.options);
// { l: 'en', psm: 6, binary: '/usr/bin/tesseract' }
var myOCR = new OCR({l:'gb', test:'abc'})
console.log(myOCR.options);
// { l: 'gb', psm: 6, binary: '/usr/bin/tesseract' }
```

Another thing to keep in mind is that this code is easily scalable, so, if we have more options in a later stage, this code won't have to be changed.

So, what do you think about this? I'm sure there must be an easier way to do what I'm doing here, but I couldn't come with anything better at the moment, let me know your thoughts!

Reply:
(PS.: Sorry for the long response hehehe)

Reply:
Hi Gabriel! Cool idea!! I did not know that was possible to delete a item in the object like that:
```
 delete options[val];
```
Awesome!!

I thought to put one more verification - check if the user put really a number where must be a number, etc...

Please take a look if I did it right :smile: 

**Before**
```
  var myOCR = new OCR({psm:'abc'})
console.log(myOCR.options);

// { l: 'por', psm: 'abc', binary: '/usr/bin/tesseract' }
```

**After**
```
function OCR(options){
  if (typeof options === "object") {
    const acceptedKeys = Object.keys(this.options);
    Object.keys(options).forEach((val) => {
      if (!acceptedKeys.includes(val)) return delete options[val];
      
         if (typeof this.options[val] !== typeof options[val])
           return delete options[val];
    });
    this.options = Object.assign(this.__proto__.options, options);
  }
}


var myOCR = new OCR({psm:'abc'})
console.log(myOCR.options);

// { l: 'por', psm: 6, binary: '/usr/bin/tesseract' }


var myOCR = new OCR({l:1, psm:'abc', binary: '/test', oi: 'tw' })
console.log(myOCR.options);

// { l: 'por', psm: 6, binary: '/test' }


var myOCR = new OCR({l:'ger', psm:'abc', binary: '/test', oi: 'tw' })
console.log(myOCR.options);

// { l: 'ger', psm: 6, binary: '/test' }

```

Reply:
Nice check, I just think we should throw a (TypeError)[https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError] in this case, what do you think?

Reply:
Cool! Even better!!